packages/core/src/editor/actions.ts

```diff
@@ -1,46 +1,43 @@
 import {
   EditorState,
-  EditorEvents,
   Indicator,
   NodeId,
   Node,
   Nodes,
   Options,
   NodeEvents,
-  SerializedNodeData,
   Tree,
+  SerializedNodes,
 } from "../interfaces";
 import {
   ERROR_INVALID_NODEID,
   ROOT_NODE,
   QueryCallbacksFor,
   ERROR_NOPARENT,
+  ERROR_ROOT_CANVAS_NO_ID,
 } from "@craftjs/utils";
 import { QueryMethods } from "./query";
+import { fromEntries } from "../utils/fromEntries";
 import { updateEventsNode } from "../utils/updateEventsNode";
 import invariant from "tiny-invariant";
-import { deserializeNode } from "../utils/deserializeNode";
-import { createElement } from "react";
-
-// TODO: move to a constants folder
-const editorEmptyState = {
-  nodes: {},
-  events: {
-    dragged: null,
-    selected: null,
-    hovered: null,
-    indicator: null,
-  },
-};
+import { editorInitialState } from "./store";
 
 export const Actions = (
   state: EditorState,
   query: QueryCallbacksFor<typeof QueryMethods>
 ) => {
   /** Helper functions */
   const addNodeToParentAtIndex = (node: Node, parent: Node, index: number) => {
-    parent.data.nodes.splice(index, 0, node.id);
-    node.data.parent = parent.id;
+    if (parent && node.data.isCanvas && !parent.data.isCanvas) {
+      invariant(node.data.props.id, ERROR_ROOT_CANVAS_NO_ID);
+      if (!parent.data._childCanvas) parent.data._childCanvas = {};
```

Comment:
Hahha okay so let's say we have the following:

```jsx
export const Card = () => {
    const { connectors: {connect, drag} } = useNode()
    return (
        <div ref={dom => connect(drag(dom))}>
            <Canvas is={CardTop} id="top"> <-- this guy
                <h1>...</h1>
            </Canvas>
            <h5>...</h5>
        </div>
    )
}

export const CardTop = () => {
    const { connectors: {connect} } = useNode()
    return (
        <div ref={connect}>
            {children}
        </div>
    )
}
```

CardTop is not a child node of Card, but it's a `_childCanvas`. Child nodes are supplied to the component's `children` prop, and we know what to render beforehand since we have the child Nodes (and so we know the type and props to render).  This however is part of the component itself. So we don't know anything about CardTop until it is actually rendered, from there we will try to get it's Node by matching its `id` to it's parent `_childCanvas` property.

```
Card_Node: {
  type: Card,
  props: {...},
  _childCanvas: {
      "top" : "CardTop_Node" // When CardTop is rendered, we find its Node by looking it up here
  }
}
CardTop_Node: {
  type: CardTop,
  props: {...}
}
```
  
(p.s. The name of `_childCanvas` can be improved, but this is addressed in the upcoming Element PR)

Reply:
thanks for the explanation! haha I do find this overall to be quite confusing, but let's tackle it in the next PR